### PR TITLE
Structured binary

### PR DESCRIPTION
**Pull request by phadej**
---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [ ] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -1,4 +1,5 @@
-packages: Cabal/ cabal-testsuite/ cabal-install/ solver-benchmarks/
+packages: Cabal/ cabal-testsuite/
+-- packages: cabal-install/ solver-benchmarks/
```
Comment (phadej) [inline cabal.project]: hups, will undo.

File: boot/SPDX.LicenseExceptionId.template.hs [hunk 2]
```diff
@@ -39,6 +40,10 @@ instance Binary LicenseExceptionId where
         then fail "Too large LicenseExceptionId tag"
         else return (toEnum (fromIntegral i))
 
+-- note: remember to bump version each time the definition changes
```
Comment (DanielG) [inline boot/SPDX.LicenseExceptionId.template.hs]: This seems to defeat the purpose of this patch. Any way this datatype can be included in the structure hash too?

Comment (phadej) [reply boot/SPDX.LicenseExceptionId.template.hs]: Manual `Binary` instance => Manual `Structured` instance.

Yet, here `Binary` instance is generated by templating, so yes, we should generate `Structured` instance too. I'll take the *license list version* and use that.

Good catch.

**Comment by phadej** [comment]
> Would be nice if there was some documentation on what exactly this is trying to do. Your commits don't have any messages, pretty hypocritical to complain when other people do this and then not have any yourself :P

This is WIP PR, I will rewrite commit messages.

**Pull request merged**